### ISSUE TITLE
set_var/remove_var are unsound in combination with C code that reads the environment

### ISSUE THREAD

**Issue by bluss** [rust-lang/rust#27970]
Like the documentation for our `set_var` (setenv) says, care must be taken with mutating environment variables in a multithreaded program. See [this glibc bug #13271](https://sourceware.org/bugzilla/show_bug.cgi?id=13271) that says getaddrinfo may call getenv.

It looks like we have an unsynchronized call to getaddrinfo and this may cause trouble with glibc/Linux.

Seeing glibc's attitude to setenv in multithreaded programs, `set_var` seems like a big hazard in general(?).

Discovered as an issue tangential to #27966

cc @alexcrichton

**Comment by bluss** [comment]
cc @luqmana

**Comment by alexcrichton** [comment]
If we did take this route I'd want to convert the environment lock to a rwlock to ensure that we could at least have parallel dns queries. Also cc #27705.

**Comment by bluss** [comment]
Yeah, either way the situation is tricky. We get tangled up in glibc internals.

**Comment by alexcrichton** [comment]
I found the wording on [this page](http://man7.org/linux/man-pages/man7/attributes.7.html) also particularly interesting:

```
       env    Functions marked with env as an MT-Safety issue access the
              environment with getenv(3) or similar, without any guards to
              ensure safety in the presence of concurrent modifications.

              We do not mark these functions as MT-Unsafe, however, because
              functions that modify the environment are all marked with
              const:env and regarded as unsafe.  Being unsafe, the latter
              are not to be called when multiple threads are running or
              asynchronous signals are enabled, and so the environment can
              be considered effectively constant in these contexts, which
              makes the former safe.
```

The getaddrinfo function [is indeed marked with this tag](http://man7.org/linux/man-pages/man3/getaddrinfo.3.html#ATTRIBUTES)

**Comment by fweimer** [comment]
`std::env::set_var` is just broken, as I commented on issue #24741:

Even without low-level races, `std::env::set_var` is not quite safe because you can change a variable (such as `TZ`), do something, and change it back to the original value, without impacting something else in the process which runs at the same time.

This problem would not go away if we provided thread-safe environment access at the libc layer. Therefore, I'm surprised the function isn't marked **`unsafe`**.

**Comment by steveklabnik** [comment]
Triage: as far as I know, this hasn't changed at all. The docs do describe this possible footgun.

**Comment by smmalis37** [comment]
This issue can also apply to any external crate that calls into something that calls get_env, for example https://github.com/time-rs/time/issues/293 and https://github.com/chronotope/chrono/issues/499. Maybe some way for these crates to observe the lock is called for?

**Comment by RalfJung** [comment]
Seems to me like this is a soundness issue, and should be marked I-unsound?

**Comment by spastorino** [comment]
Assigning `P-medium` as [discussed as part of the Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/I-prioritize.20.2327970.20Does.20getaddrinfo.20need.20to.20take.20the.20ENV.20l.E2.80.A6) and removing `I-prioritize`.

**Comment by jhpratt** [comment]
Just repeating what I've said on Zulip: To resolve the soundness bug in time & chrono, one option would be to provide a `std::env::lock` method, which holds a lock on the environment until dropped. It's my understanding this is basically what is happening internally on some functions already in stdlib. Exposing this would allow _much_ safer usage of FFI.

**Comment by RalfJung** [comment]
The env lock has to be exposed somehow, I agree.

An alternative API would be a function that takes a closure that will be called with the lock held. This is less powerful (which might be a good thing: provide the least powerful API that is good enough).

**Comment by RalfJung** [comment]
However, talking of the env lock -- also see https://github.com/rust-lang/rust/issues/64718 for how the current use of the env lock is wrong, so exposing it in that state might be a problem.

**Comment by jhpratt** [comment]
Sure, that would work as well. End result is the same, if done properly.

**Comment by DemiMarie** [comment]
Honestly, I don’t believe `std::env::set_var` can be made safe.

In a pure-Rust program, it is already safe, but in C, `setenv` is considered thread-unsafe.  Therefore, C libraries assume that they can call `getenv` at any time.  Furthermore, adding a call to `getenv` is probably not considered worthy of a major version bump in the C library, so what was previously a safe call into that library can become unsafe without warning!

Even without the thread safety problems, there are other reasons that `std::env::set_var` is unsafe.  Libraries such as `gtk` will load plugins from a path specified in an environment variable, and loading a shared library is inherently unsafe.  If the shared library is in the system shared library directory, it can be assumed safe, since the OS is considered trusted.  But it isn’t safe to (say) load a library from the Downloads folder (which could contain anything).

What *might* be safe is a function that fails if multiple threads are running, and which only allows setting a limited number of env vars, such as the timezone.

**Comment by CraftSpider** [comment]
I think that while the thread unsafety is a significant issue, security unsafety is not the purpose of `unsafe`, and it's not practical for the std to decide what variables count as 'safe' sec-wise, as that is heavily dependent on both platform and even installed programs.

**Comment by DemiMarie** [comment]
What I meant is that while one could reasonably argue that writing into `/proc/self/mem` is unsafe, writing into `~/Downloads` is fairly clearly safe.  But if one is allowed to set GTK’s plugin path, then it is impossible to write a safe GTK binding.

This doesn’t even consider C libraries that call `getenv` from background threads.  I believe this is considered perfectly okay in C, but `std::env::set_var` being safe makes writing safe bindings to such libraries impossible.  There is no way (other than disgusting linker hacks) to make them synchronize their calls to `getenv` with Rust’s calls to `setenv`, so either we give up binding to such libraries, or we declare `std::env::set_var` unsafe.  Not being able to bind to such libraries safely would be a shame, IMO.  An unsafe `std::env::set_var` would be much less of a loss.

**Comment by ssokolow** [comment]
I think it makes more sense to call it "out of scope".

Setting environment variables is such a common thing in interacting with the greater software ecosystem that I think declaring it unsafe would be akin to re-legislating whether leaking memory (and, thus, `Box::leak` and using `Rc` and `Arc` in ways that can't be proven free of reference cycles) should be `unsafe`.

**Comment by carbotaniuman** [comment]
I don't know why there's a tangent about security unsafety, but I think the parent was talking about how set_env automatically races with every single use of getenv by C/C++/FFI code. A data race is as UB as it gets.

**Comment by ssokolow** [comment]
My point is that `set_env` has so many uses in purely safe projects that marking it unsafe would be like a more minor version of marking `println!` and friends unsafe because an FFI dependency might write to stdout or stderr without acquiring the lock... and I doubt you'd consider that a reasonable change.

It honestly wouldn't surprise me if such a change resulted in a `safed_setenv` crate that just restores the old status quo (with a clear explanation of what it's doing and why in the README) so that `set_env` isn't the one and only thing keeping an otherwise pure safe Rust crate from using `#![forbid(unsafe_code)]`.

**Comment by carbotaniuman** [comment]
I understand that unsafe should only be used for memory safety, and that's what I'm saying. Glibc, musl, basically all Unix libcs says that messing with setenv is out of scope, and their official reocmmnedation/requirement is to not call it when you have threads.

I'm mainly talking about the `localtime_r` and `getaddrinfo` case. If the only way to express those functions **soundly** is to be unsafe with the caveat that `set_var` has not been called, maybe we put the emphasis on the wrong with.

**Comment by Nemo157** [comment]
`std::env::set_var` could be made sound to FFI by making it purely affect the Rust runtime's environment variables. That means any variable set by it could only be read via `std::env::get_var`, and would probably make it more expensive as it would have to CoW the C environment variables, but that is the only approach that seems to allow disconnecting its soundness from the soundness of the underlying `setenv` call.

**Comment by DemiMarie** [comment]
Another approach would be to have `std::env::set_var` fail if multiple threads are running.  This is trivial on Linux if `/proc` is mounted.

This is my preferred approach for several reasons:

- It allows programs that call `std::env::set_var` early in `main` to continue to work.  This is useful for setting parameters for C libraries, for example.
- We can expose the “are multiple threads running” function in from libstd.  This is useful for calling inherently thread-unsafe C functions such as `curl_global_init`, which otherwise cannot be safely called from Rust.

**Comment by jhpratt** [comment]
@DemiMarie By fail, I presume you mean panic?; `env::set_var` doesn't return a `Result`/`Option`. It's certainly an option, though the behavior would certainly have to be clearly documented.

<sup>(for those wondering, the uptick in activity is because I created a thread on IRLO)</sup>

**Comment by DemiMarie** [comment]
@jhpratt yes, panic.  I can’t think of any legitimate uses for changing environment variables (which, by definition, are global state) when multiple threads are running.

**Comment by ssokolow** [comment]
@DemiMarie I'd be in favour of doing that and adding a couple of other versions, following the pattern of many other panicking operations in `std`:

* `fn set_var_checked` (Returns a `Result` with the error variant covering "threading in use" but also also taking over the existing "may panic if `key` is empty, contains an ASCII equals sign `'='` or the NUL character `'\0'`, or when the value contains the NUL character")
* `unsafe fn set_var_unchecked` (Used for situations where threads are at play, but a manual audit has confirmed that only one of them has anything to do with the environment.)

**Comment by DemiMarie** [comment]
@ssokolow I like `set_var_checked`, but prefer to avoid `set_var_unchecked`.  The problem is that `set_var_unchecked` doesn’t seem to have useful semantics: even if *right now* no other thread accesses the environment, this could change in a patch version bump of a dependency.  I cannot think of any legitimate uses for it right now, and I suspect that those that do exist would be okay with calling `setenv` manually.

**Comment by ssokolow** [comment]
@DemiMarie Fair enough. I suppose it's rare enough to need to change in-process environment after the initial setup phase that asking people adding a dependency on `libc` or `winapi` would be acceptable when one of the following is insufficient:

* Set the environment before starting the threads
* Use `Command::env*` methods when invoking subprocesses
* Do everything single-threaded

**Comment by joshtriplett** [comment]
Following up on this:

@RalfJung wrote:
> The env lock has to be exposed somehow, I agree.

I don't think we should expose the internals of our environment locking, because that means we'll never be able to change or evolve it again.

I looked into using `getaddrinfo_a` with `GAI_WAIT`, since it's marked as "MT-Safe" rather than "MT-Safe env locale" like `getaddrinfo` is, but I think this is a [bug in the documentation](https://lore.kernel.org/linux-man/YEZwqXhBTzbMAyzo@localhost/T/#u), because `getaddrinfo_a` still calls the same `getaddrinfo` in a thread, so it'll still fail in the same way if the environment changes concurrently.

For the moment at least, I do think the right solution is to take the environment lock around calls to `getaddrinfo`. This doesn't solve the general problem, since non-Rust code won't use the Rust environment lock, but it'll at least mean purely Rust programs won't have any issues here, which is an improvement over the current situation.

(And yes, we should also switch that lock to an rwlock. There were some efforts to do that recently, which [ran into an issue that required reverting them](https://github.com/rust-lang/rust/pull/82877), but I think it'd be possible to make the same change again as long as we avoid releasing a lock in a child process that we acquire in the parent. Once we do that, switching it to an rwlock seems fine, and will avoid a performance bottleneck on `getaddrinfo` calls.)

Ideally we should have a better solution for the environment in the future, but in the meantime, using the environment lock here seems like a good idea.

**Comment by jhpratt** [comment]
@joshtriplett What would you consider exposing the internals of the environment locking? In the same comment, Ralf threw out the option of a closure that would be run with the lock held. That wouldn't result in any public API exposing the implementation detail, but would still result in the same behavior.

**Comment by DemiMarie** [comment]
@joshtriplett What about having our environment-modification functions `panic!` if there is more than one thread running?  That seems to be the only sustainable approach in the long run.

**Comment by joshtriplett** [comment]
@jhpratt That would hide the exact mechanism we use for locking, but would still expose that we use locking around the environment. It would also likely expose semantics of mutual exclusion. For instance, does it support recursive locking or can it self-deadlock? Does it use mutual exclusion (in which case people could count on two such closures not running concurrently), or a read lock, or some other mechanism such as snapshotting or transactions? What if we have a target where the environment is multi-thread safe; can we just call the closure with no locking, or would that break assumptions made based on other targets?

This is an area that still needs some fixing, and I'd be concerned that exposing its internals will make it much harder for us to improve it in the future.

@DemiMarie That would break a huge number of programs that work today. Rust's locking makes those programs correct as long as they only use Rust functions to touch the environment.

**Comment by DemiMarie** [comment]
> @DemiMarie That would break a huge number of programs that work today. Rust's locking makes those programs correct as long as they only use Rust functions to touch the environment.

The problem is that there are far, *far* too many ways to touch the environment.  A C library that a Rust program depends on could start touching the environment in a patch release, and without warning.  Furthermore, there is currently no way to safely call C libraries that read environment variables.  See https://github.com/rust-lang/rust/issues/27970#issuecomment-725873403 for cases where this is a problem.  Rust bindings to GTK would also be affected.  And even if a C library does not currently access the environment, it might start doing so in a minor version.  If I maintained such a library, I would not consider adding a call to `getenv` to be a breaking change.

In short, on \*nix, calling `setenv` or `putenv` (or assigning to `environ`) from a multi-threaded program is *itself* Undefined Behavior, unless one knows every C function the other threads could possibly be calling.  And that is not a realistic assumption in practice.  That leaves panicking as the only option.  Is it a good one?  No.  But it is the best one available to us.   And it unlocks critical functionality such as proper time zone handling.

**Comment by jhpratt** [comment]
> That would break a huge number of programs that work today. Rust's locking makes those programs correct as long as they only use Rust functions to touch the environment.

The problem is that's not always possible. If it were, I likely wouldn't even be subscribed to this issue, as there wouldn't have been a soundness issue in the time crate.

**Comment by DemiMarie** [comment]
> > That would break a huge number of programs that work today. Rust's locking makes those programs correct as long as they only use Rust functions to touch the environment.
> 
> The problem is that's not always possible. If it were, I likely wouldn't even be subscribed to this issue, as there wouldn't have been a soundness issue in the time crate.

Exactly :+1:

**Comment by joshtriplett** [comment]
I'm not arguing that the general state of environment handling in threaded programs is anything less than awful.

It *is* worth pointing out that glibc, at least, seems to have its own internal locking, such that concurrent calls to `setenv` and `getenv` are not *inherently* racy. And since Rust calls those functions internally rather than reimplementing environment handling itself, the combination of Rust and glibc shouldn't in general have races. (There are exceptions, notably around fork/exec, and those need further work.)

The issue is that today, there are many, many programs out there that use threads, use Rust environment functions, and either don't call C or don't call any C code that touches the environment. Those programs run fine today, and will break if we change our environment functions to panic.

I understand why, if you've run into a problem caused by this, panicking if called from a multithreaded program seems like it might be the best option. But I don't think the breakage that would cause is something we should do lightly. And I'm still hoping there are better alternatives.

**Comment by DemiMarie** [comment]
> I'm not arguing that the general state of environment handling in threaded programs is anything less than awful.
> 
> It _is_ worth pointing out that glibc, at least, seems to have its own internal locking, such that concurrent calls to `setenv` and `getenv` are not _inherently_ racy. And since Rust calls those functions internally rather than reimplementing environment handling itself, the combination of Rust and glibc shouldn't in general have races. (There are exceptions, notably around fork/exec, and those need further work.)

The glibc documentation states otherwise.  So does POSIX.  What we *might* be able to do is atomically replace `extern char **environ` with a different pointer, leaking the old one.  But even that is doubtful: it could race with libc, which won’t access `environ` atomically.  And it will leak memory.

> The issue is that today, there are many, many programs out there that use threads, use Rust environment functions, and either don't call C or don't call any C code that touches the environment. Those programs run fine today, and will break if we change our environment functions to panic.
> 
> I understand why, if you've run into a problem caused by this, panicking if called from a multithreaded program seems like it might be the best option. But I don't think the breakage that would cause is something we should do lightly. And I'm still hoping there are better alternatives.

I don’t like the idea of panicking, but from what I can tell, that is the only option available if we want `std::env::{set,remove}_var` to actually modify the C environment.  Of course, if we are willing for those functions to only affect the environment as seen by Rust, then everything is fine.  Using `setenv` from a multithreaded process isn’t, unless one controls all of the threads in the process.

Also, do you know of any specific programs that *do* call `std::env::{set,remove}_var` with multiple threads running?  Those functions are usually called during initialization.  The main exception I can think of is programs that use PAM, but PAM is inherently unsafe in a multithreaded process.

**Comment by jhpratt** [comment]
Out of curiosity, what's the chances of getting a crater run if there were a speculative PR put together changing `setenv` to panic when multiple threads are running?

Note that I don't particularly care for this solution, but it _is_ a solution. If fallout is minimal, why not do it? If there's significant fallout, at least we can eliminate it.

**Comment by DemiMarie** [comment]
> Out of curiosity, what's the chances of getting a crater run if there were a speculative PR put together changing `setenv` to panic when multiple threads are running?
> 
> Note that I don't particularly care for this solution, but it _is_ a solution. If fallout is minimal, why not do it? If there's significant fallout, at least we can eliminate it.

This actually seems like the best answer to me.  IMO we should really do the same thing with `std::process::exit`; which has similar problems.

**Comment by jhpratt** [comment]
Yeah, that's definitely true :/ How would a "real world case study" as you mention work? Aside from just asking people if they ever run `setenv` when multiple threads are running, which would likely produce very skewed results.

**Comment by joshtriplett** [comment]
Offhand, one common example would be running your main function inside an async runtime that starts threads, and then setting environment variables at the top of that async main.

**Comment by jhpratt** [comment]
This is me mostly speaking of ignorance: What are real-world use cases for setenv? I've never used it. If I understood _where_ and _why_ it's used, I can think about potential workarounds and their respective tradeoffs a bit.

The async runtime is a reasonable concern. While one could certainly write out the macro expansion for an async main, it would require some thought from the programmer's perspective.

**Comment by leo60228** [comment]
It's not very reliable, but here's a reproducer: https://gist.github.com/d3709b9b3a4b5338ef89c98b28ad078e

On my machine (24 cores), you can run `while :; do timeout 0.1 target/release/gai-crash; done` and get a segfault within a few seconds.

**Comment by saethlin** [comment]
> Furthermore, there is currently no way to safely call C libraries that read environment variables. See #27970 (comment) for cases where this is a problem.

I think this comment understates the problem. openssl calls `getenv` during certificate validation, because it calls `gmtime_r`. If the community is going to take this family of problems seriously, I think the `chrono`/`time` advisories are just the start.

**Comment by thomcc** [comment]
There's significant (and ongoing) discussion about this https://internals.rust-lang.org/t/synchronized-ffi-access-to-posix-environment-variable-functions/15475.

**Comment by saethlin** [comment]
I totally agree, that would not be a good outcome. Thanks for directing me to the discussion, I'm not used to looking at IRLO.

**Comment by fweimer** [comment]
> I think this comment understates the problem. openssl calls `getenv` during certificate validation, because it calls `gmtime_r`*. If the community is going to take this family of problems seriously, I think the `chrono`/`time` advisories are just the start.
> 
> * well, it calls `getenv` if linked against glibc, and `gmtime_r` is `MT-Safe env` so any implementation may but I don't think musl does.

We should be able to make the *internal* `getenv` calls in glibc MT-safe at least. The external interface is a different matter. POSIX even hints towards a thread-safe implementation for the public `getenv` function (*The returned string pointer might also be invalidated if the calling thread is terminated.*), although this would be an observable difference to the C11 `getenv`.

**Comment by DemiMarie** [comment]
@fweimer would it be possible to provide a way for Rust to detect this at runtime?

**Comment by DemiMarie** [comment]
@fweimer Would it be desirable to:

1. Issue an advisory for this, so that it gets backported by distribution maintainers, and
2. Provide an interface that Rust can use at runtime to detect a patched glibc?

@RichFelker would you be interested in changing this in musl?

**Comment by leo60228** [comment]
This issue doesn't solely apply to glibc and musl. We'd need to ensure this for every target, or get it changed in the standard. I don't think either option is feasible.

**Comment by saethlin** [comment]
Patching POSIX/glibc has been proposed in the IRLO thread linked above. I don't think it's a good solution because this takes the ability to fix things out of the hands of the Rust programmers, and in many cases leaves us at the behest of whatever distro we happen to be using (there is at least one comment to this effect in the thread). And in addition, this does nothing to help with programs or library that access the environment through `environ` (this is mentioned multiple times in the IRLO thread).

Moreover, I don't think it's at all fair to put this on libc maintainers. Rust messed up here. Or, if not the standard library, loads of library authors messed up. And even if they did, I can't imagine prioritizing this as a backported security fix. IMO it's a pretty clear case of someone didn't follow the very old and well-documented interface. I can't imagine that if this had been brought up around 1.0 that all of `std::env` would look like what it does today. To that end: https://github.com/rust-lang/rust/pull/92365

---

It would be _nice_ if POSIX or libc authors decided to make `getenv` and `setenv` both thread-safe, because it would mitigate the impact of this kind of mistake, as it has been made outside of Rust too: https://sourceware.org/bugzilla/show_bug.cgi?id=15607#c3

**Comment by fweimer** [comment]
> @fweimer Would it be desirable to:
> 
>     1. Issue an advisory for this, so that it gets backported by distribution maintainers, and
> 
>     2. Provide an interface that Rust can use at runtime to detect a patched glibc?

We don't have patches yet. It's unclear how backportable they would be.

This is definitely *not* a glibc security vulnerability (if that's why mean by “advisory”). `setenv` is not required to be thread-safe.

**Comment by richfelker** [comment]
> @richfelker would you be interested in changing this in musl?

No. At has been explained above and in the other tracker item of this form I took part in, this is purely a bug in the Rust specification, and needs to be fixed there. `setenv` (the POSIX function) is not thread-safe, and for every good reason should not be required to be. "Not thread-safe" means you are essentially not allowed to use it in a multi-threaded program at all. (There may be some subtle exceptions to that, but it's a good enough approximation.)

Why is it important that `setenv` not be thread-safe? Because the interface for accessing the environment, `getenv`, provides the caller a pointer to storage whose lifetime could end when `setenv` is called, but the caller of `getenv` has no way to "release" its reference to this storage. The other sanctioned interface, accessing `environ[]` directly, would have data races under concurrent `setenv`, which is just as bad. So if `setenv` were thread-safe, the operations of *reading* the environment would necessarily be non-thread-safe. Present issues of POSIX do not require that `getenv` be thread-safe, but basically everything depends on it being thread-safe, and all implementations I'm aware of make it so (which requires no work, since it's accessing read-only data). And at least for implementation-internal access to the environment, it's a requirement that it be thread-safe. And future issues of POSIX [will require](https://www.austingroupbugs.net/view.php?id=188#c325) it to be.

For what it's worth, `getaddrinfo` is not one of the functions which POSIX permits to inspect the environment, and the musl implementation does not, so this particular manifestation of the Rust issue should not be one you encounter on musl.

**Comment by briansmith** [comment]
I don't think it's reasonable to change the behavior of the current `set_var` to panic or otherwise fail because that's a backward-incompatible change. Nor do I think it's reasonable to add any locking overhead or complexity to `get_var`.

What makes sense to me is to:
* Add an `unsafe fn set_var_unchecked(...)` that does exactly what the current set_var does.
* Clearly document the thread safety hazard in `set_var_unchecked`.
* Change `set_var` so that it is a wrapper around `set_var_unchecked`.
* Mark `set_var` as `#[deprecated = "use set_var_unchecked"]`.
* Expand the documentation of `set_var` to explain why it is deprecated.
* In the next edition, make `set_var` an `unsafe fn`.

That's it. No added locking, no additional checking of `/proc`, no trying to patch libc. Leave `get_env` and similar as regular safe functions; they *are* safe as long as nobody does an unsafe thing using `set_var/set_var_unchecked`.

@ssokolow wrote:
> There are entire design philosophies for microservices which call for you to pass all your configuration parameters via environment variables instead of files or command-line arguments
For that kind of application, can't you just use `Command::{env, envs}` so that the environment variables are set when the process is spawned, instead of using `set_env`?

@ssokolow wrote:
> My point is that set_env has so many uses in purely safe projects

I don't think there has been evidence to support this claim yet.

> There are a lot of projects that one can find that use set_var in its test suite. That's not necessarily a big problem for Rust in the large, that's just a scaffolding issue

It is an issue for those tests because they unsafe because `set_env` actually is unsafe. Even if `set_env` were somehow safe, such tests are often racing with each other (FWIW, I found this issue trying to find a solution for projects who had tests misbehaving due to `set_env`).

**Comment by ssokolow** [comment]
Based on the discussions here and on IRO, I've since come around to the view that marking it unsafe in a way that doesn't break existing code is the right thing to do, as long as it's understood that it's not "because of FFI data races".

(i.e. That it should be `unsafe` but it's important to think of the reasoning in terms of "In day to day coding, we generally don't think of `std`'s dependency on `libc` as FFI, but this this can make safe Rust fail because of it". It's not something Rust can fix because of libc-internal uses, and it's something that has the potential to bite you even if your program uses only safe Rust and no third-party crates which do FFI.)

**Comment by jhpratt** [comment]
On the contrary I believe that this should have an unsafe equivalent (in addition to the safe version being deprecated) because of FFI. As someone who calls a libc method that obtains an environment variable, I uphold all safety invariants of that method. It is std that is not upholding the safety invariants of set_env. Yes, it potentially manifests in pure Rust code, but it's much more likely when dealing with FFI. Rust upholding a safety invariant does not mean upholding it only when used with other Rust code. FFI is just as important.

**Comment by ssokolow** [comment]
> On the contrary I believe that this should have an unsafe equivalent (in addition to the safe version being deprecated) because of FFI.

I'm not sure who you're replying to. Both I and @briansmith argued for this functionality to exist in `unsafe` form.

**Comment by jhpratt** [comment]
I was addressing this specific part

> as long as it's understood that it's not "because of FFI data races".

Though I'm certain we'd agree that we don't need to agree on the _reason_ for doing this so long as it's done.

**Comment by ssokolow** [comment]
Ahh. My point there was that, originally, I was seeing it more as "`LD_PRELOAD` can break Rust safety invariants, as can things like GTK plugin loading, so why are we complicating safe Rust code that relies only on the FFI in `std`?"

When you "turn to FFI", there's an expectation that either you or your dependency developer will take on the responsibility of abstracting away what unsafety can be abstracted and documenting what is an inherently unsafe aspect of what you're depending on.

`std` technically uses FFI, but to think about it the same way would fatally undermine what Rust is trying to accomplish as far as safety goes.

**Comment by RalfJung** [comment]
> Nor do I think it's reasonable to add any locking overhead or complexity to get_var.

The current get_var already performs locking; I don't think anyone proposed to add more locking.
Under your proposal, one could consider removing that locking, though probably it is better to keep it for compatibility reasons? OTOH, removing that lock would finally fix https://github.com/rust-lang/rust/issues/64718.

**Comment by briansmith** [comment]
> The current get_var already performs locking; I don't think anyone proposed to add more locking. 

OK, I didn't realize that. This doesn't change my proposal. The `set_var_unchecked` that I am proposing would continue to do the locking, since I defined it to do exactly what `set_var` does now.

> Under your proposal, one could consider removing that locking, though probably it is better to keep it for compatibility reasons? OTOH, removing that lock would finally fix #64718 (or at least the part of it that is about the env lock).

Since the locking versions are already implemented, I suspect some people already rely on that locking in their pure-Rust (or close enough) code, especially tests. In #64718, there was already a plan proposed that seems to fix that issue without having to get rid of the lock?

**Comment by saethlin** [comment]
> I suspect some people already rely on that locking in their pure-Rust (or close enough) code, especially tests

Yes. We're aware of ~300 published crates which rely on this https://crater-reports.s3.amazonaws.com/pr-90799/index.html

**Comment by comex** [comment]
@richfelker

> "Not thread-safe" means you are essentially not allowed to use it in a multi-threaded program at all. (There may be some subtle exceptions to that, but it's a good enough approximation.)

The problem is that most people don't know about this rule and so their code doesn't follow it, even in otherwise high-quality codebases.  As a test, I hacked up glibc to log to syslog whenever `setenv` is called with multiple threads running, then installed this in a Debian VM and started up some random desktop sessions.  Here's what I got:

https://gist.github.com/comex/9fa0899293d1c6e748faa81b875b25ac

It's inevitable that Rust code will end up in the same address space as code that does this, whether it's a Rust binary using C libraries or vice versa.  But this leads to the extreme conclusion that portable Rust code cannot safely access the environment *at all*, not even just to read it.  It can't even safely copy the environment at initialization time, not if built into a library that might be `dlopen`ed by C code while threads are already running.  

Perhaps all that code should be fixed.  (After all, some of these instances probably represent dormant race conditions in the C code itself, even without Rust coming into the picture.)  But the above link represents only a small sample of Linux desktop applications, and Linux desktop applications represent a tiny fraction of the environments people run Rust code in.  I suspect that unchecked `setenv` use can be found throughout the entire ecosystem of multithreaded C programs (and [Python programs](https://bugs.python.org/issue39375), and [Ruby programs](https://marc.info/?l=llvm-dev&m=136942524817592&w=2), et cetera).  Fixing it all is impossible.

The alternative, changing multiple libcs for multiple operating systems by multiple vendors that don't cooperate on anything these days is… also a tall order, to be sure.  Even so, it seems like a more viable path.  @zackw, a contributor to glibc, has helped get the ball rolling.  He has a draft proposal for how to make the environment thread-safe without breaking compatibility too badly.  I don't think he has gotten around to revising it and inviting "feedback from a wider community" yet, but since this conversation is happening now, I'll link it in case you're interested:

https://internals.rust-lang.org/t/synchronized-ffi-access-to-posix-environment-variable-functions/15475/178

**Comment by richfelker** [comment]
>  But this leads to the extreme conclusion that portable Rust code cannot safely access the environment at all, not even just to read it. It can't even safely copy the environment at initialization time, not if built into a library that might be dlopened by C code while threads are already running.

This "conclusion" makes no sense at all. Any C code doing this has **extreme** **serious** memory-corruption race-condition bugs, including when used from C, and needs to be fixed. It is not a thread-safe or a library-safe operation and you would not expect to find it in library code. "Rust becomes unsafe if you `dlopen` C code with extreme memory-safety violations" is not a surprise or a problem for Rust to try to mitigate. If you `dlopen` C code that does `*(char *)rand() = 42;` you get to keep both pieces, and this is no different.

**Comment by RalfJung** [comment]
> How many of those are multi-threaded programs, and how many of them are doing the setenv/putenv anywhere other than at startup or after fork before exec'ing something?

Quoting comex, emphasis mine: "print a message to syslog whenever setenv or putenv is called **while multiple threads [are] running**"

**Comment by richfelker** [comment]
OK. From the look of the types of vars, it still appears these are being set to pass to a child process. I wonder if the evaluation of the "multiple threads" predicate is wrong after fork. I suppose I should dig in and look at one or more of them. Obvious approach would be just setting breakpoints on `setenv` and `putenv`.

**Comment by DemiMarie** [comment]
> No it is not. The interface contract **fundamentally does not admit any "fix"** for this.

Is leaking memory an option?

**Comment by leo60228** [comment]
Even if it works in most cases, an atomic write and non-atomic read is still a data race, right?

**Comment by DemiMarie** [comment]
Yes

**Comment by comex** [comment]
> OK. From the look of the types of vars, it still appears these are being set to pass to a child process. I wonder if the evaluation of the "multiple threads" predicate is wrong after fork. I suppose I should dig in and look at one or more of them. Obvious approach would be just setting breakpoints on `setenv` and `putenv`.

It does correctly handle fork.  In case anyone wants to reproduce the results, [here is a slightly less ugly version](https://gist.github.com/comex/d80d45e05379e08769df5abe838d5528) of the glibc patch I used.  It can also capture core dumps.

> No it is not. The interface contract **fundamentally does not admit any "fix"** for this.

It is unfixable in some cases, namely:
1. Accessing `environ` directly;
2. Calling `putenv` on a buffer, and then modifying the buffer.

The draft proposal I mentioned proposes to deprecate these, since they're relatively rare.  For more typical use cases, the interface can be 'fixed' (well, band-aided over) by ensuring that environment variable buffers are never freed.  It also proposes adding new functions to cover use cases that currently require accessing `environ` directly, such as listing variables.

Yes, it's all a big hack.  But I am pretty convinced that the alternatives are worse.

**Comment by tarcieri** [comment]
> What makes sense to me is to:
> 
> * Add an `unsafe fn set_var_unchecked(...)` that does exactly what the current set_var does.
> * Clearly document the thread safety hazard in `set_var_unchecked`.
> * Change `set_var` so that it is a wrapper around `set_var_unchecked`.
> * Mark `set_var` as `#[deprecated = "use set_var_unchecked"]`.
> * Expand the documentation of `set_var` to explain why it is deprecated.
> * In the next edition, make `set_var` an `unsafe fn`.

For what it's worth, according to a straw poll on this thread, an overwhelming majority of people preferred this approach:

https://internals.rust-lang.org/t/synchronized-ffi-access-to-posix-environment-variable-functions/15475/155

**Comment by briansmith** [comment]
The current documentation of `env::set_var` says: "Note that while concurrent access to environment variables is safe in Rust, some platforms only expose inherently unsafe non-threadsafe APIs for inspecting the environment. As a result, extra care needs to be taken when auditing calls to unsafe external FFI functions to ensure that any external environment accesses are properly synchronized with accesses in Rust." This is misleading. Rust code that calls `env::{set_var,get_var}` is unsafe even if it never calls any unsafe functions, if it is in a process (e.g. Firefox) that uses a different mechanism (e.g. a different lock) to protect access to the environment. So, at a minimum we should fix this documentation right away.

Also, I'd like to add an additional suggestion to my proposal:
* In the updated documentation for these functions, encourage people to use `Command::{env,envs}` when spawning processes instead of `set_env`, whenever practical.

**Comment by thomcc** [comment]
I think those steps are not particularly uncontroversial (and I'm happy to see the concensus[^1]), the open question is whether or not that's sufficient.

[^1]: I'm *very* happy to see it being taken seriously, in a way that isn't "expose libstd's lock" (which would be a problem for no_std code) and "forbid calling functions that read from the environment" (which would be a problem for FFI code). I was pretty worried we'd land on one of those for quite a while...

Concretely, I voted "yes" on that poll, agree that the listed steps are a good idea, but... think it isn't sufficient for several types of existing users to be able to migrate in many cases.

I worry that without a migration path to a safe function, users will throw up their hands and either `#[allow(deprecated)]` or use `unsafe { set_var_unchecked(...) }`, without being able to guarantee it's sound.

**Comment by RalfJung** [comment]
> Let's use Firefox as an example. Firefox has, through NSPR, its own safe way to use the environment using locks. Rust has its own. But, they aren't safe to use together unless they agree to use the same lock. I.e. Firefox should be patching libstd to use the NSPR lock or patching NSPR to use the Rust lock, if it wants non-unsafe Rust code to be able to safely read and write the environment. The same with StarOffice, etc. This patching is probably not happening, so in general Rust libraries must avoid calling setenv/env::set_var completely, if they want to be thread-safe and be safe to use in such applications.

It's worse than that: Rust libraries must avoid calling `env::get_var` because that could race with a `setenv` from NSPR! Likewise, Firefox must not link with any C library that calls `getenv`, since that call would also fail to be protected by the lock.

The NSPR lock is just as bad as the Rust lock when it comes to compositionality with general-purpose C code. Both of them only work if they are used by 100% of the code that reads or mutates the environment. How does Firefox avoid the `getaddrinfo` problem that started this thread? The `getenv` inside `getaddrinfo` is just as unsafe in Firefox as it is in a Rust FFI situation.

**Comment by briansmith** [comment]
Regarding the specific issue of whether libstd's `getaddrinfo` wrapper should have a lock around the call to the C function, see https://emptysqua.re/blog/mac-python-getaddrinfo-queueing/, where the author notes that lock contention was a serious problem with CPython's implementation of exactly that idea. Rust's env lock would need to become, at minimum, a reader-writer lock so that the `getaddrinfo` calls do not get serialized in a high-performance multithreaded application. Even then, I think most applications that would want to call `getaddrinfo` and which care about performance would accept the overhead of acquiring a read lock.

**Comment by tarcieri** [comment]
> Rust's env lock would need to become, at minimum, a reader-writer lock

It already is: 

https://github.com/rust-lang/rust/blob/5e93f6e318e687c05c8c44517de4586ed75ce3f4/library/std/src/sys/unix/os.rs#L487

```rust
static ENV_LOCK: StaticRWLock = StaticRWLock::new();
```

**Comment by briansmith** [comment]
> It's worse than that: Rust libraries used by Firefox must avoid calling `env::get_var` because that could race with a `setenv` from NSPR! Likewise, Firefox must not link with any C library that calls `getenv`, since that call would also fail to be protected by the lock.

That's not how Rust's memory unsafety model works, right? The unsafe action is the call to `setenv` in a thread executing concurrently with any other thread that might call `getenv`. Just like any thread could use `unsafe` code to scribble on memory concurrently with a thread running safe Rust code, and cause the safe Rust code to fault.

> The NSPR lock is just as bad as the Rust lock when it comes to compositionality with general-purpose C code. Both of them only work if they are used by 100% of the code that reads or mutates the environment. How does Firefox avoid the `getaddrinfo` problem that started this thread? The `getenv` inside `getaddrinfo` or `localtime_r` is just as unsafe in Firefox as it is in a Rust FFI situation.

Generally environment variables aren't to be used in Firefox because it has a different configuration mechanism. Also, Firefox has many bugs, and I did see some that are still open related to memory safety and the environment.

On non-Windows targets Firefox does appear to use NSPR `PR_GetAddrInfoByName`, and `PR_GetAddrInfoByName` does seem to have a lock around `getaddrinfo` in some circumstances. I'm not sure whether it uses the lock on Linux since it seems to be an autoconf-type mess. Also, Firefox's acceptable performance characteristics aren't necessarily every applications' acceptable performance characteristics.

**Comment by briansmith** [comment]
> On non-Windows targets Firefox does appear to use NSPR `PR_GetAddrInfoByName`, and `PR_GetAddrInfoByName` does seem to have a lock around `getaddrinfo` in some circumstances. I'm not sure whether it uses the lock on Linux since it seems to be an autoconf-type mess. Also, Firefox's acceptable performance characteristics aren't necessarily every applications' acceptable performance characteristics.

Oh, that lock in `PR_GetAddrInfoByName` separate from the NSPR environment lock, and is to guard against non-reentrant getaddrinfo. So, Firefox may indeed have a memory safety bug here, if `getaddrinfo` reads from the environment around the time a `PR_SetEnv`/etc. is called in another thread.

**Comment by RalfJung** [comment]
> That's not how Rust's memory unsafety model works, right? The unsafe action is the call to setenv in a thread executing concurrently with any other thread that might call getenv. Just like any thread could use unsafe code to scribble on memory concurrently with a thread running safe Rust code, and cause the safe Rust code to fault.

I am not quite sure what you mean by "Rust's memory unsafety model", but if a Rust function uses `get_var` and, concurrently, another thread uses the NSPR-protected `setenv`, then we have a data race and hence UB. This is the exact same problem that makes a call to `getaddrinfo` UB when there is a concurrent Rust call to `set_var`: we end up calling `setenv` and `getenv` concurrently without (proper) synchronization, which POSIX says is UB and which can lead to data races in the actual implementations.

**Comment by RalfJung** [comment]
As @richfelker keeps saying: under current POSIX, the only correct thing to do in a concurrent program is to never call `setenv` once the first thread spawns. Adding a lock to protect environment accesses is *not* a viable alternative unless you can guarantee that the same lock protects *all* environment accesses. We are having this discussion because the Rust lock can not actually make this guarantee. I don't think Firefox's NSPR lock can make the guarantee, either -- unless Firefox patches every single library it links with (be it written in C, Rust, or whatever) to ensure all environment accesses use the NSPR lock.

Having a lock clearly doesn't work for Rust, that's why we are here. Why would you think that it works any better for Firefox/NSPR?

**Comment by briansmith** [comment]
> I am not quite sure what you mean by "Rust's memory unsafety model", but if a Rust function uses `get_var` and, concurrently, another thread uses the NSPR-protected `setenv`, then we have a data race and hence UB.

`setenv` is an unsafe function, and is breaking invariants that Rust doesn't enforce. This is equivalent to `*(char *)rand() = 42` that @richfelker noted above.

>  This is the exact same problem that makes a call to `getaddrinfo` UB when there is a concurrent Rust call to `set_var`: we end up calling `setenv` and `getenv` concurrently without (proper) synchronization, which POSIX says is UB and which can lead to data races in the actual implementations.

That's because `env::set_var` is an unsafe function and should be deprecated and eventually made `unsafe`.

> Having a lock clearly doesn't work for Rust, that's why we are here. Why would you think that it works any better for Firefox/NSPR?

I hope nothing I've written above makes it seem like I'm advocating for adding a lock or using the lock more. My proposal is in https://github.com/rust-lang/rust/issues/27970#issuecomment-1003218748.

**Comment by RalfJung** [comment]
> I hope nothing I've written above makes it seem like I'm advocating for adding a lock or using the lock more.

@briansmith you wrote
> Thus, just because that multi-threaded program is calling setenv doesn't necessarily mean it is doing so unsafely; these may be the "subtle exceptions" that @richfelker was referring to.

I hope the subsequent discussion established that Firefox's lock is *not* one of these "subtle exceptions". It is just as broken as Rust's lock.

**Comment by briansmith** [comment]
> unless Firefox patches every single library it links with (be it written in C, Rust, or whatever) to ensure all environment accesses use the NSPR lock.

FWIW, this is indeed what Firefox developers are expected to do. 1. Remove as many uses of the environment as possible. 2. Change any that can't be removed to use `PR_SetEnv`/`PR_GetEnv`. Note that it seems the Firefox developers overlooked the fact that `getaddrinfo` may access the environment; perhaps they were relying on it conforming to POSIX. If I were still working on Firefox I would add to the backlog the task of removing all the `PR_SetEnv` calls.

**Comment by briansmith** [comment]
> So, we are back to @comex' point that the current POSIX API is broken in the sense that it does not fit the need of the ecosystem

I agree.

> -- the ecosystem needs a thread-safe way to access _and mutate_ the environment

A very small part of the ecosystem needs that.

> At the same time there is no way to fix this issue anywhere other than in libc, since there is no other way to ensure that all libraries linked into a process use the same lock.

There are multiple issues:
1. Currently, `env::set_var` is an unsafe function, but it isn't marked `unsafe`. This is a serious bug that should be fixed immediately. I provided a proposal for this above.
2. `getaddrinfo` on Linux (in glibc at least) accesses the environment even though it isn't supposed to, which is problematic if anything concurrently modifies the environment. What should we do about it? I propose that we don't try to paper over it, and just make sure all the APIs we provide for modifying the environment are marked `unsafe`, and document the hazards clearly.
3. It would be nice to have a non-`unsafe` version of `env::set_var` that is actually safe to use. In general, this seems to require cooperation from libc, as you and @comex note. This would be a decades-long process and it isn't very urgent for us to deal with, IMO.

**Comment by RalfJung** [comment]
> There are multiple issues:

I agree on all 3 of these, including your approach for (1) -- except I am not sure if changing `unsafe`ty of a function at an edition boundary is possible (and if it is, there are some other candidates, such as `Command::before_exec`), but that is mostly independent of the rest of the plan.

**Comment by briansmith** [comment]
Besides the three issues I noted above, there is a fourth issue: `getenv` is also not thread-safe:

@richfelker wrote:
> And future issues of POSIX [will require](https://www.austingroupbugs.net/view.php?id=188#c325) it to be.

This is the change in that doc:

> At line 33878 change:
> 
>     The getenv() function is inherently not thread-safe because it
>     returns a value pointing to static data.
> 
> to:
> 
> Some earlier versions of this standard did not require getenv()
> to be thread-safe because it was allowed to return a value
> pointing to an internal buffer. However, this behaviour allowed
> by the C Standard is no longer allowed by POSIX.1. POSIX.1
> requires the environment data to be available through environ[],
> so there is no reason why getenv() can't return a pointer to the
> actual data instead of a copy. Therefore getenv() is now required
> to be thread-safe (except when another thread modifies the
> environment).

We could either keep using `getenv` on the assumption that even though current and older POSIX says it isn't thread-safe, or we could just change `std::env` to use `environ` directly with pure Rust code and avoid `getenv()` completely; i.e. RiiR. I prefer to RiiR, though I think it should be a lower priority.

**Comment by richfelker** [comment]
@briansmith wrote:

> Thus, just because that multi-threaded program is calling `setenv` doesn't necessarily mean it is doing so unsafely; these may be the "subtle exceptions" that @richfelker was referring to.

It does mean it's doing so unsafely. The only "subtle exception" is that the unsafety is defined as making the call concurrently with any other standard function, so a program where the only other thread(s) do not call library functions at all, only do raw computation and communication of results via raw atomics without invoking synchronization primitives, could in theory be well-defined while calling thread-unsafe functions fron a single thread. In reality, such programs don't exist, and they're not relevant to the topic at hand. For all practical purposes, if your program is multithreaded at the time of a call to `setenv`, it has UB.

In particular, putting locks around a call to a non-thread-safe function (and whatever other functions you expect "might access the same internal state it does") does not make the function thread-safe. If it's non-thread-safe, you just can't use it.

**Comment by richfelker** [comment]
This isn't a response to any one comment, but I'd like to address a general theme I've seen that the environment could somehow be "made thread-safe for modification", either with a different API (more powerful than `getenv` that gives the caller a way to express the lifetime of its reference to the string) or by leaking memory (note: this would be bad for legacy single-threaded apps that repeatedly change `TZ` or similar, which are actually a thing). This is at best pointless, because what you're asking for is making it "safe" to change string-valued global variables out from under code using them. Even if you could change the value of a variable safely in a way that doesn't involve machine-level data races, there's still a race between setting the value and using the value where another thread could set it to something different. It's possible that some or even most of these *happen to be avoided* by virtue of threads having different roles where they don't set or use the same environment variables as each other, but it's a fundamentally unsafe programming model, and we've known it's unsafe for something like half a century.

The only legitimate role the environment has in a safe programming model is as immutable input context at program invocation, where it acts similar to `argv[]` but implicitly inherited-unless-modified rather than specific to a particular program invocation. And for this it has always worked very well.

**Comment by briansmith** [comment]
To add onto what @richfelker said, there's no way to implement a thread-safe `env::set_var` that modifies the environment returned by `getenv`, because `getenv` must use the environment in `environ[]` and because `environ` is a global variable and not a function, there's no way to add a lock to it transparently.

**Comment by tarcieri** [comment]
> The only legitimate role the environment has in a safe programming model is as immutable input context at program invocation, where it acts similar to argv[] but implicitly inherited-unless-modified rather than specific to a particular program invocation. And for this it has always worked very well.

As long as we're spitballing about hypothetical new libc features, how about a feature that opts-in to actually enforcing this, i.e. a way to `freezeenv` the environment within a particular process such that calls to `setenv`/`unsetenv` return an error at the libc level?

If that were possible, it seems like the safety errors around `getenv` would just go away without the need for synchronization (besides a process-wide "is environment frozen" atomic flag).

Of course, instead the `setenv`/`unsetenv` callers would have to deal with a runtime error, but based on the above conversation it sounds like there's no truly sound solution to fix these safety issues without ensuring that the callers simply don't use these functions inside of programs where multiple threads access the environment.

**Comment by saethlin** [comment]
> I worry that without a migration path to a safe function, users will throw up their hands and either `#[allow(deprecated)]`

I don't want to harp on this well-known situation too much: `itoa` used `mem::uninitialized` for 2 years after it was deprecated for being essentially impossible to call, and for sure the way `itoa` used it was UB. Migration might be slower than we would hope.

**Comment by DemiMarie** [comment]
> If it's non-thread-safe, you just can't use it.

I wonder if libc should expose an `is_single_threaded` function.  If this returns true, mutating the environment is safe.  The main use for this would be in assertions, such as:

```c
int main(int argc, char *argv[]) {
    some_struct *s = some_initialization_that_might_use_threads();
    if (!is_single_threaded()) {
        fputs("BUG: some_initialization_that_might_use_threads didn’t stop the threads it created!\n");
        abort();
    }
    modify_environment(s);
    rest_of_program();
}
```

> The only legitimate role the environment has in a safe programming model is as immutable input context at program invocation, where it acts similar to `argv[]` but implicitly inherited-unless-modified rather than specific to a particular program invocation. And for this it has always worked very well.

This also means that any library that accepts configuration via the environment, without allowing this configuration to be overridden in a thread-safe way, is significantly lacking, if not outright buggy.  If one must use such libraries, the only option available is to call `setenv`, `putenv`, or `unsetenv` early in `main`, while the program is single-threaded.

It is also worth noting that one cannot modify the environment in a child process after `fork()`, if the parent process is multi-threaded.  POSIX only allows async-signal-safe interfaces to be used there, which `*env` are not.  So the only safe approach I am aware of is to allocate an entirely new environment block in the parent process and use `execve`.

**Comment by comex** [comment]
> This is at best pointless, because what you're asking for is making it "safe" to change string-valued global variables out from under code using them. Even if you could change the value of a variable safely in a way that doesn't involve machine-level data races, there's still a race between setting the value and using the value where another thread could set it to something different. It's possible that some or even most of these _happen to be avoided_ by virtue of threads having different roles where they don't set or use the same environment variables as each other, but it's a fundamentally unsafe programming model, and we've known it's unsafe for something like half a century.

Arguably this is an example of a difference between C and Rust philosophy.  You're treating memory safety as a special case of correctness, reasoning that there is no point making it memory-safe if it can't be correct.  From my perspective, an API that encourages logical race conditions is like a house that's not built to code, while a memory-unsafe API is like a house that's \*\*on fire\*\*.

To be fair, I was suggesting changes in lib\*c\*.  And if Rust needs a temporary workaround anyway, given that all changes to OS components take a long time to become ubiquitous, then perhaps there is no use in seeking a fix at an intermediate level between 'temporary workaround to put out the fire' (whatever that may be) and 'long-term solution' (moving away from the environment entirely).

It almost seems like the best workaround would be a tiny library to provide a thread-safe shadow environment, which could hypothetically be shared by all the codebases that currently put locks around setenv, ranging from Rust's libstd to Mozilla's codebase to Python, etc.  That way, at least those codebases would be safe to use together.  Unfortunately, various features of the ecosystem make adding a dependency far more costly than just the size of the code itself, so such an approach is probably not realistic.

**Comment by RalfJung** [comment]
> In particular, putting locks around a call to a non-thread-safe function (and whatever other functions you expect "might access the same internal state it does") does not make the function thread-safe. If it's non-thread-safe, you just can't use it.

POSIX explicitly allows calling `MT-Unsafe const:X` functions if suitable locking is in place on *all* concurrent calls that are marked with `MT-(Un)Safe X`. So putting locks around these *can* make them thread safe, the difficulty is in ensuring that the locks are put *everywhere*. That can only be done in a "closed world" situation where one has full control over all code linked into the process.

> Even if you could change the value of a variable safely in a way that doesn't involve machine-level data races, there's still a race between setting the value and using the value where another thread could set it to something different. It's possible that some or even most of these happen to be avoided by virtue of threads having different roles where they don't set or use the same environment variables as each other, but it's a fundamentally unsafe programming model, and we've known it's unsafe for something like half a century.

It's not a great programming model for sure, but it is not "unsafe" in the same sense that triggering UB is unsafe. (It is not `unsafe` in the Rust sense, in particular.) And it can be perfectly fine when different variables are used in different threads. Unfortunately POSIX acts as a bug amplifier where such code that might have a logic bug or might even be fine is then viewed as having full UB. That's a footgun if I've ever seen one.

> The only legitimate role the environment has in a safe programming model is as immutable input context at program invocation, where it acts similar to argv[] but implicitly inherited-unless-modified rather than specific to a particular program invocation. And for this it has always worked very well.

Then please remove `setenv` and `unsetenv` entirely. But as long as those functions exist, the reality of this API contradicts your statement. POSIX clearly intends the environment to be mutable. Or rather, it lures programmers by pretending that the environment can be mutated, only to hit them with a club if they try to actually mutate it. This is the API design equivalent of an ambush.

Okay, maybe I got carried away a bit here. ;) But I really don't think it is fair to blame programmers here.

**Comment by briansmith** [comment]
> It almost seems like the best workaround would be a tiny library to provide a thread-safe shadow environment, which could hypothetically be shared by all the codebases that currently put locks around setenv, ranging from Rust's libstd to Mozilla's codebase to Python, etc. That way, at least those codebases would be safe to use together. Unfortunately, various features of the ecosystem make adding a dependency far more costly than just the size of the code itself, so such an approach is probably not realistic.

You could imagine defining an API like this, or a better API:
```
void shared_env_read_lock() __attribute__((weak));
void shared_env_read_unlock() __attribute__((weak));
void shared_env_write_lock() __attribute__((weak));
void shared_env_write_unlock() __attribute__((weak));
```
Each standard library (Ruby, NSPR, Rust libstd) would provide an implementation of that API and it wouldn't matter which implementation the linker chose as long as it chose the same implementation for all 4 functions. Even libc could opt into doing it. This works poorly for `getaddrinfo` and any other function that does I/O, which is back to the original problem here. For `getaddrinfo`, i wouldn't feel great about holding even a read lock over I/O, though I could maybe be convinced that it is harmless enough considering nobody is supposed to be calling any of the functions that would acquire the write lock anyway.

Before that though, I'd like to try harder to eradicate uses of `set_var` in the Rust ecosystem, since I think many library authors have likely overlooked the case where they are being used in a process that isn't pure Rust and where the non-Rust code is using a different lock to protect the environment or not using a lock at all. I think the process of trying to eradicate `set_var` will teach us what primitives we'd need to add. In particular, I suspect we'll get to the point where the next step is to expand the testing harness to allow safely running a test with environment variables set to particular values.

**Comment by briansmith** [comment]
A long time ago @thomcc wrote:
> but probably the solution is probably not to report a CVE against nearly every rust library that does FFI.

I think it doesn't make sense to file bugs on projects that are just reading the environment. However, I do recommend filing issues against crates that are modifying the environment by calling `env::set_ver` or `env::remove_var` and making PRs to help them find an alternate solution (hopefully `Command::{env,envs}`). I would suggest prioritizing non-test instances over test instances.

**Comment by richfelker** [comment]
@DemiMarie wrote:

> This also means that any library that accepts configuration via the environment, without allowing this configuration to be overridden in a thread-safe way, is significantly lacking, if not outright buggy.

Yes, absolutely.

> It is also worth noting that one cannot modify the environment in a child process after `fork()`, if the parent process is multi-threaded. POSIX only allows async-signal-safe interfaces to be used there, which `*env` are not. So the only safe approach I am aware of is to allocate an entirely new environment block in the parent process and use `execve`.

If you need to modify the environment in the child without exec'ing, you can modify `environ` directly. However there are very few if any (likely no) applications for this since the functions which would consume it are almost surely AS-unsafe. Note however that relaxation of the rule about what you can do after MT `fork` is a common extension implementations offer, albeit inconsistently. (For example glibc goes to the trouble of making sure `malloc` works here, but doesn't explicitly try to make anything else work, and what does vs doesn't work is undocumented but widely relied upon, including in some language runtimes like Ruby where the unwritten contract spilled into the language's contract for programs written in the language...)

**Comment by richfelker** [comment]
@RalfJung wrote:

> POSIX explicitly allows calling `MT-Unsafe const:X` functions if suitable locking is in place on _all_ concurrent calls that are marked with `MT-(Un)Safe X`. So putting locks around these _can_ make them thread safe, the difficulty is in ensuring that the locks are put _everywhere_. That can only be done in a "closed world" situation where one has full control over all code linked into the process.

This is not the case. `MT-Unsafe const:X` is a glibc invention that does not match the POSIX definition of thread-safety at all, and that's the source of a lot of this confusion. These glibc annotations were invented to describe the implementation properties of glibc and make what was previously a de facto contract with glibc into a documented one.

The [POSIX definition of thread-safe](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_407) is:

> A thread-safe function can be safely invoked concurrently with other calls to the same function, or with calls to any other thread-safe functions, by multiple threads. Each function defined in the System Interfaces volume of POSIX.1-2017 is thread-safe unless explicitly stated otherwise. Examples are any "pure" function, a function which holds a mutex locked while it is accessing static storage, or objects shared among threads.

Thread-unsafe is not separately defined, but in the absence of anything more specific, the only reasonable interpretation is that functions which are not thread-safe are not required to have these properties. In particular you cannot assume thread-unsafe functions "can be safely invoked concurrently... with calls to any other thread-safe functions".

In particular, there is no mention of different classes/categories "X" of thread-unsafety.

Thread-safety specifically in regards to the environment is also touched on in [2.9.1](https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_09_01):

> Since multi-threaded applications are not allowed to use the environ variable to access or modify any environment variable while any other thread is concurrently modifying any environment variable, any function dependent on any environment variable is not thread-safe if another thread is modifying the environment; see XSH exec.

This text is redundant with the thread-unsafety of `setenv`, but also covers the case where `environ` is modified directly.

**Comment by nico-abram** [comment]
If `std::env::set_var` is deprecated and/or made unsafe, should something like a safe `std::os::windows::env::set_var` be added to provide a migration path to users of non POSIX platforms?

**Comment by briansmith** [comment]
From the glibc official documentation (https://www.gnu.org/software/libc/manual/html_node/Environment-Access.html#Environment-Access):
> Libraries should use secure_getenv instead of getenv, so that they do not accidentally use untrusted environment variables. 

I just filed https://github.com/rust-lang/rust/issues/92558 about that.

> Modifications of environment variables are not allowed in multi-threaded programs. The getenv and secure_getenv functions can be safely used in multi-threaded programs.

@richfelker Where can we find the similar documentation for musl? Specifically, I'm looking for documentation that says `getenv` is thread safe in the same way that glibc says it is.

@nico-abram wrote:
> If std::env::set_var is deprecated and/or made unsafe, should something like a safe std::os::windows::env::set_var be added to provide a migration path to users of non POSIX platforms?

That sounds reasonable to me, though I think it could be done as an independent follow-up.

**Comment by richfelker** [comment]
@briansmith wrote:

> > Libraries should use secure_getenv instead of getenv, so that they do not accidentally use untrusted environment variables.
> 
> I just filed #92558 about that.

This is an oversimplification and really depends on what they're doing with the variables.


> > Modifications of environment variables are not allowed in multi-threaded programs. The getenv and secure_getenv functions can be safely used in multi-threaded programs.
> 
> @richfelker Where can we find the similar documentation for musl? Specifically, I'm looking for documentation that says `getenv` is thread safe in the same way that glibc says it is.

Presently there is no such documentation. Having documentation for guarantees beyond the ones the standard makes is an open todo item. For this specific case, however, the POSIX-next requirement serves as the documentation.

**Comment by briansmith** [comment]
Sorry about all the discussion about deprecating `env::{set_var,remove_var}` above. I didn't realize there was a separate issue about it: #90308. I'll direct all my future comments about the deprecation there.

**Comment by RalfJung** [comment]
> This is not the case. MT-Unsafe const:X is a glibc invention that does not match the POSIX definition of thread-safety at all, and that's the source of a lot of this confusion. These glibc annotations were invented to describe the implementation properties of glibc and make what was previously a de facto contract with glibc into a documented one.

I stand corrected, sorry for that -- I did not realize that this was all glibc-specific.

> I mostly wouldn't be opposed to that, but that isn't how standards that document & establish consensus on existing practice work.

Existing practice clearly involves calling `setenv` in multi-threaded programs, so it does not seem to me like POSIX is interested in accounting for existing practice.

**Comment by richfelker** [comment]
@RalfJung wrote:

> Existing practice clearly involves calling `setenv` in multi-threaded programs, so it does not seem to me like POSIX is interested in accounting for existing practice.

Here existing practice means what existing implementations support as well as what existing applications are using from that set. Bugs in existing applications that are going unnoticed or willfully unfixed aren't part of said "existing practice".

**Comment by comex** [comment]
> > Since multi-threaded applications are not allowed to use the environ variable to access or modify any environment variable while any other thread is concurrently modifying any environment variable, any function dependent on any environment variable is not thread-safe if another thread is modifying the environment; see XSH exec.
> 
> This text is redundant with the thread-unsafety of `setenv`, but also covers the case where `environ` is modified directly.

Notably, this implies that it *is* safe for a multithreaded program to access the environment under a lock, as long as it does so by accessing `environ` directly rather than using `getenv` or `setenv`, and as long as no other threads simultaneously call "any function dependent on any environment variable".  However, it's not clear to me if there are any guarantees around what functions are "dependent on any environment variable", e.g. whether glibc's `getaddrinfo` behavior is conforming.  Quite possibly it is, in which case this guarantee is fairly useless.

**Comment by richfelker** [comment]
@comex wrote:

> However, it's not clear to me if there are any guarantees around what functions are "dependent on any environment variable", e.g. whether glibc's `getaddrinfo` behavior is conforming. Quite possibly it is, in which case this guarantee is fairly useless.

The standard is very explicit where functions are specified to use the environment, so my interpretation is that this behavior is non-conforming, and that's why we don't do it in musl. If I recall, we've had this discussion with glibc folks a few times in the past and the consensus was that stdlib extensions using the environment where it's not specified really should be scanning the environment at program entry time to extract the settings they want to use, so that the specified safety properties are maintained. I think this is the path glibc is following for adding any new tuning variables, though the existing ones (at least for now) continue to behave as they always have.

**Comment by tmccombs** [comment]
> The standard is very explicit where functions are specified to use the environment

But does it explicitly state that functions cannot add additional dependencies on environment variables that are not explicitly specified?

**Comment by evanj** [comment]
From the issue description: "may cause trouble with glibc/Linux": Here is a Rust program that only uses the standard library and crashes in Linux with glibc with a Segmentation Fault. It calls `ToSocketAddrs` on one thread, which calls `getaddrinfo`, and calls `set_env` on another. https://github.com/evanj/cgogetenvcrash/blob/main/rustsetenvcrash/src/main.rs

**Comment by VorpalBlade** [comment]
What is the status of this? Specifically with regards to exposing the env lock to users of std, this was talked about in 2020 but nothing much seems to have happened? Was the idea dropped, if so where is the rationale for that documented?

**Comment by ChrisDenton** [comment]
glibc says calling set_env is unsound in the presence of threads. So in order to address this we need the application to essentially become single threaded while set_env is being called.

The environment lock falls short on this. We can't seriously mandate that everyone acquires a lock before making libc calls, not least because the lock is Rust only.

A more robust course of action would be to simply (and more accurately) mark set_env unsafe and teach how it can be safely used in a cross-platform manner. In this scenario the lock is basically redundant so it could even be considered for removal.

**Comment by tarcieri** [comment]
I believe this is the current plan:

https://github.com/rust-lang/rust/pull/116888#issuecomment-1852620311

> We discussed this in the libs-api meeting. What we would like to see happen is that this function gets turned into an unsafe function for the 2024 edition, but on older editions only emits a warning if not inside an unsafe block (possibly using deprecated_safe https://github.com/rust-lang/rust/issues/94978).

**Comment by richfelker** [comment]
Plan sounds good.

The right thing for applications to do is *not modify the environment*. If they're modifying it to pass to child processes, the right way to do that is passing an explicit new environment, not modifying your own environment and relying on it getting copied. If they're using the environment to control the behavior of C library code they're calling, that's inherently unsafe and the library should be fixed to accept explicit caller-provided overrides of whatever it was getting from the environment.

**Comment by VorpalBlade** [comment]
> Plan sounds good.
> 
> The right thing for applications to do is _not modify the environment_. If they're modifying it to pass to child processes, the right way to do that is passing an explicit new environment, not modifying your own environment and relying on it getting copied. If they're using the environment to control the behavior of C library code they're calling, that's inherently unsafe and the library should be fixed to accept explicit caller-provided overrides of whatever it was getting from the environment.

@richfelker This doesn't really work if you are implementing a shell though, unless you decide to emulate your internal environment completely (and pass that on to tasks you spawn. That is a niche use case sure, and I'm curious as to what nushell and fish (which was recently rewritten in Rust) do here.

**Comment by tbu-** [comment]
> unless you decide to emulate your internal environment completely (and pass that on to tasks you spawn).

Note that "emulating the internal environment completely" essentially only means keeping a hashmap around and accessing that instead of `std::env::var`.

> I'm curious as to what nushell and fish (which was recently rewritten in Rust) do here.

fish doesn't call `std::env::set_var`: https://github.com/search?q=repo%3Afish-shell%2Ffish-shell+set_var&type=code
nushell doesn't call `std::env::set_var` outside of test code: https://github.com/search?q=repo%3Anushell%2Fnushell+set_var&type=code.

**Comment by richfelker** [comment]
> unless you decide to emulate your internal environment completely (and pass that on to tasks you spawn).

Of course this is the correct way to do a shell. Just because ancient unix history did things the sloppy way of modifying the shell process's environment directly rather than keeping its own data structures for key/value map doesn't mean a modern one written in a modern safe language should copy the same bad ideas. It's not even easier to store exports in your own environment, since you already need your own map for non-exported variables, and you can just use the same one (with export flag on each entry) for both.

**Comment by tmccombs** [comment]
> If they're using the environment to control the behavior of C library code they're calling, that's inherently unsafe and the library should be fixed to accept explicit caller-provided overrides of whatever it was getting from the environment.

It is really unfortunate that libc and posix have functions that depend on environment variables, such as `localtime_r` and anything that uses the locale.  I really wish the c and posix standard would make a more multithreading friendly API.

**Comment by richfelker** [comment]
There is never a reason to set the locale environment vars within a program. If you want to request a specific locale, you pass the name to the locale functions. The environment is only there as a default inherited at execution time.

For time zones, indeed the standard library functionality is rather poor. There have been proposals to modernize it with first class zone objects, but not a lot of progress on making it happen.

**Comment by kamulos** [comment]
Just for completeness sake: there is also https://github.com/sunfishcode/eyra#why which solves the issue on the level of the libc

**Comment by DemiMarie** [comment]
> For time zones, indeed the standard library functionality is rather poor. There have been proposals to modernize it with first class zone objects, but not a lot of progress on making it happen.

Is this something that could be prototyped in a C library?

**Comment by tmccombs** [comment]
Sure. It isn't trivial though. On linux, you need to know how to read the timezone files, which is kind of complicated to do, I think that is what chrono does though. I'm not sure if there is a documented and supported way to do it on other major platforms. I think Mac uses the same tzinfo format, but I'm not sure if that is documented to remain stable. I don't know about windows, or mobile oses.

**Comment by briansmith** [comment]
> I've posted a patch that should fix this for glibc:

Just to clarify, your patch doesn't fix it "enough" for Rust to consider them to `set_var`/`remove_var` to be safe, because C code is still allowed to read directly from `environ` and do other things that aren't protected by your improvements. That doesn't mean your improvement isn't welcome though.

**Comment by richfelker** [comment]
I don't see how this provides any significant thread-safety. The core unsafety of modifying the environment is that the lifetime of a string returned by `getenv` can end asynchronously if another thread changes the environment. The only way to avoid this is to leak everything ever added.

**Comment by tbu-** [comment]
Another way is for `getenv` to put the result into a thread-local variable. This is compatible with POSIX's guarantees, but probably not in line with glibc's backward compatibility guarantees.

**Comment by RalfJung** [comment]
From the patch:

> I think this achieves full thread safety and async-signal-safety for
getenv if the environment is only updated using setenv, unsetenv,
clearenv.  If putenv is used, it depends on how the passed string is
updated afterwards.  Likewise for direct environ writes.  This is fairly
easy to accomplish because the current implementation never frees any
variables set using setenv.

So, seems to me like setenv is already leaking, and this patch builds on that.

Very nice to see progress on this!

**Comment by VorpalBlade** [comment]
Even if that patch is accepted in glibc, POSIX doesn't guarantee thread safety, so this cannot change the decision for Rust (unless you get a change into musl, all the *BSDs, OpenSolaris, Darwin/Mac OS X etc, as well as the POSIX standard itself). And I don't foresee that happening due to backward compatibility around `environ` as well as the issue mentioned with `putenv`.

Still, it is a step in the right direction.

**Comment by tmccombs** [comment]
Well, if it weren't for direct access to environ and the putenv issue, there could possibly be a safe function that is only available on platforms that use a new enough version of glibc.

**Comment by tbu-** [comment]
> Even if that patch is accepted in glibc, POSIX doesn't guarantee thread safety, so this cannot change the decision for Rust (unless you get a change into musl, all the *BSDs, OpenSolaris, Darwin/Mac OS X etc, as well as the POSIX standard itself).

Note that Rust doesn't strictly adhere to POSIX, Rust is fine with stuff that all the platforms implement. E.g. `pthread_t` is assumed to be an integer even though POSIX explicitly says that it might not be.

I haven't looked into whether musl, *BSDs or Darwin implement synchronization for environment variables. I just wanted to say that POSIX itself might not be a requirement.

**Comment by tmccombs** [comment]
What if another thread reads environ directly? Would that still be undefined behavior unless setenv uses an atomic operation to swap the pointer, *and* the reader uses an atomic load to read environ?

**Comment by tmccombs** [comment]
My point is that even if a c library reads from environ directly, then no amount of synchronization around setenv in rust would be safe in a multi-threaded program.

**Comment by tgross35** [comment]
> I've posted a patch that should fix this for glibc:
> 
>     * [[PATCH] stdlib: Make getenv thread-safe in more cases](https://inbox.sourceware.org/libc-alpha/87le2od4xh.fsf@oldenburg.str.redhat.com/)
> 
> 
> Reviews would be appreciated. It does not solve the conceptual problem that the environment is a process-global resource. Other threads still observe temporary environment changes intended for the current thread only. Just the crashes and missed variable lookups are gone.

A version of this has since been committed https://github.com/bminor/glibc/commit/7a61e7f557a97ab597d6fca5e2d1f13f65685c61.
